icefall/context_graph.py

```diff
@@ -0,0 +1,222 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from typing import Dict, List, Tuple
+
+
+class ContextState:
+    """The state in ContextGraph"""
+
+    def __init__(self, token: int, score: float, total_score: float, is_end: bool):
+        """Create a ContextState.
+
+        Args:
+          token:
+            The token id.
+          score:
+            The bonus for each token during decoding, which will hopefully
+            boost the token up to survive beam search.
+          total_score:
+            The accumulated bonus from root of graph to current node, it will be
+            used to calculate the score for fail arc.
+          is_end:
+            True if current token is the end of a context.
+        """
+        self.token = token
+        self.score = score
+        self.total_score = total_score
+        self.is_end = is_end
+        self.next = {}
+        self.fail = None
+
+
+class ContextGraph:
+    """The ContextGraph is modified from Aho-Corasick which is mainly
+    a Trie with a fail arc for each node.
+    See https://en.wikipedia.org/wiki/Aho%E2%80%93Corasick_algorithm for more details
+    of Aho-Corasick algorithm.
+
+    A ContextGraph contains some words / phrases that we expect to boost their
+    scores during decoding. If the substring of a decoded sequence matches the word / phrase
+    in the ContextGraph, we will give the decoded sequence a bonus to make it survive
+    beam search.
+    """
+
+    def __init__(self, context_score: float):
+        """Initialize a ContextGraph with the given ``context_score``.
+
+        A root node will be created (**NOTE:** the token of root is hardcoded to -1).
+
+        Args:
+          context_score:
+            The bonus score for each token(note: NOT for each word/phrase, it means longer
+            word/phrase will have larger bonus score, they have to be matched though).
+        """
+        self.context_score = context_score
+        self.root = ContextState(token=-1, score=0, total_score=0, is_end=False)
+        self.root.fail = self.root
+
+    def _fill_fail(self):
+        """This function fills the fail arc for each trie node, it can be computed
+        in linear time by performing a breadth-first search starting from the root.
+        See https://en.wikipedia.org/wiki/Aho%E2%80%93Corasick_algorithm for the
+        details of the algorithm.
+        """
+        queue = []
+        for token, node in self.root.next.items():
+            node.fail = self.root
+            queue.append(node)
+        while queue:
+            current_node = queue.pop(0)
+            current_fail = current_node.fail
+            for token, node in current_node.next.items():
+                fail = current_fail
```

Comment:
Another possible pitfall, my feeling is that there should be some sort of recursive(or looping) fall-back here. For example, to fill the failure arc of C, you might look at its parent B, and possibly also B's failure node A, and so on...

Reply:
It is strange the experiment results keep the same after fixing the `fill_fail`. I think the phrases are complex enough (it has 800 phrases) to make some difference. There indeed are some bugs in the code, I will try visualizing the constructed graph tomorrow.

Reply:
> Another possible pitfall, my feeling is that there should be some sort of recursive(or looping) fall-back here. For example, to fill the failure arc of C, you might look at its parent B, and possibly also B's failure node A, and so on...

Yes, you are right.